src/crypto/sha256sum.cpp

```diff
@@ -0,0 +1,117 @@
+#include "crypto/sha256sum.h"
+
+#include <algorithm>       // min
+#include <cstdio>          // BUFSIZ
+#include <fstream>
+#include <ios>             // ios_base::{binary, in, ate}
+#include <openssl/evp.h>
+
+#include "misc_log_ex.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "util"
+
+namespace {
+	constexpr size_t SHA256_MD_SIZE = 32; // 256/8
+}
+
+namespace tools
+{
+	bool sha256sum(const uint8_t *data, size_t len, crypto::hash &hash)
+	{
+		// Check hash size
+		if (sizeof(hash.data) < SHA256_MD_SIZE) {
+			MERROR("crypto::hash instance passed to sha256sum() is too small: only " << sizeof(hash.data) << " bytes");
+			return false;
+		}
+
+		unsigned int final_digest_size;
+		if (!EVP_Digest(data, len, (unsigned char*) hash.data, &final_digest_size, EVP_sha256(), NULL)) {
+			MERROR("Failed to digest with OpenSSL SHA-256 EVP");
+			return false;
+		} else if (final_digest_size > sizeof(hash.data)) { // EVP_Digest overflows buffer. shouldn't ever happen, but you never know
```

Comment:
These size checks also seem a bit too paranoid to me. This only goes wrong if you're not actually using a SHA256 algorithm, no? Which would imply a pretty broken OpenSSL build.

Reply:
Yeah, they're pretty paranoid I'd say. In practice, these would only trigger if the OpenSSL implementation was broken. IIUC though, officially, up to `EVP_MAX_MD_SIZE` (36 bytes I believe)  can be written from a call to `EVP_DigestFinal_ex` which is larger than `HASH_SIZE`.

Reply:
If you asked for SHA256 and got 36 bytes back you didn't get what you asked for. I think we should be allowed to expect that our library dependencies are actually doing what we expect them to.

Reply:
That's a good point, I guess we've got other things to worry about if OpenSSL is broken. I pushed a new commit with your suggestions